### PR TITLE
Implemented logic gates

### LINKED ISSUE #25: When the player starts a game, the game engine should display a logic gate
## Summary:
When the player selects "Challenge Mode" or "Practice Mode" from the main screen, the game should display a single logic gate.

## Criteria:

- [x] Player must be able to see the logic gate
- [x] The logic gate must include two inputs going in, presenting the problem to solve
- [x] Player must be able to solve the logic gate
	- The logic gate must include an area for user input

## Unresolved questions:
- What is the process to generate the logic gates?

## Possible Solutions:
- You could create the logic gates in the JavaScript file using objects. For further information, please look at [this specification ticket](https://github.com/the-new-developers/gatehopper/issues/9).

## Additional notes:
- For now, we just want the simplest solution - we can complicate it later. It might be best if it just shows the same logic gate every time; any logic you implement around random generation will likely be overwritten later, when we try and implement procedural generation.

### LINKED ISSUE #44: Implement boolean logic gates.
Logic gates that need to be added
- [x] NOT
- [x] OR
- [x] AND
- [x] XOR
- [x] NOR
- [x] NAND
- [x] XNOR

It probably goes without saying that they should connect.

### PR DESCRIPTION
**Pull request by VictorSohier**
## Changes:
Closes #25 
- Gates are displayed
- Tint is for testing purposes
- User input may be handled by invisible image buttons

Closes #44 
- All logic gates implemented

## How to test this PR:
Go to challengeMode.js and swap out this.gate to change behavior. Changing the values between 1 and 0 (or true and false if you prefer) for constant gates changes "input" values to the gate.

The image input is such that the images may be pre-scaled. Show scaled may benefit from caching the scaled output for future use.

## Checklist:
- [x] tested locally
- [x] added tests if needed
- [x] linked the associated ticket
- [x] assigned yourself to this PR

(Side note, this means we can make a computer within a computer)

### CODE REVIEW

File: src/objects/gate.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+		if (this.constructor === Gate)
+		{
+			throw new TypeError("Abstract class \"Gate\" cannot be instantiated directly.");
+		}
+		this.img = img;
+	}
+
+	answer()
+	{
+		if (this.schema === undefined)
```
Comment (rodneybarnes) [inline src/objects/gate.js]: I'm not able to see where `schema` is being set/should be set - is this a property on all js classes or something?

File: src/objects/xnorGate.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import { XorGate } from "xorGate.js";
```
Comment (rodneybarnes) [inline src/objects/xnorGate.js]: I tried swapping in the XnorGate here and got a TypeError in the console:

```
index.html:1 Uncaught TypeError: Failed to resolve module specifier "xorGate.js". Relative references must start with either "/", "./", or "../".
```

When I added a './' to the file path here, I then got a similar error for the import statement in `xorGate.js` for `twoInputGate.js`, and when I fixed that one I saw a separate error related to how that class was being imported:

```
Uncaught SyntaxError: The requested module './twoInputGate.js' does not provide an export named 'TwoInputGate'
```
I removed the curly braces around the import in `xorGate.js` and saw an error in the way I was importing `xnorGate.js` in the challenge mode scene.

I'd recommend going through these classes to make sure they all `export default class`, which will make importing them easier (so we don't have to use named imports for some and not others), and then also testing to make sure they're all wired up correctly (i.e. that the file paths are correct).

File: src/scenes/challengeMode.js [hunk 3]
```diff
@@ -1,4 +1,30 @@
...
+export default class ChallengeMode extends Scene {
+    constructor() {
+        super();
+    }
+
+    setup() {
+    	this.constant1;
+    	this.constant2;
+    	this.gate;
+    	let img = loadImage("assets/textures/gates/No Input Buffer.png");
```
Comment (rodneybarnes) [inline src/scenes/challengeMode.js]: Is it possible to have each gate we might `new` up here load its own image somehow? It would be nice to be able to call `new AndGate` and it can take care of loading its own image without the dev needing to know which png file to use every time.

**Review by rodneybarnes — COMMENTED** [review]
Tested locally and the gates show up great - I left a comment about testing the import statements for the other gates to make sure they all work and fixing any that don't.

I like how we're using inheritance, however personally I find our object hierarchy very complex and somewhat confusing. On the one hand, it makes a lot of sense to say that the `answer` for a `xnor` gate is the opposite of the answer for a `xor` gate, however should these objects get any more complex I think the coupling will become an issue and trying to debug them will be challenging as well (for example it took me a bit to unravel all the errors I was getting around the import/export issues I commented on).

I think that refactoring towards a simpler hierarchy would make it easier for other devs to work with, though since the complexity here isn't an immediate issue I think it's safe to leave this as-is. Would you be able to provide some documentation on how to use this? For instance, it looks like the `ConstantGate` is used to represent player input?

Because we're using vanilla Javascript here we don't have a nice type system letting us know what parameters like `input1` and `img` are, so we'll have to augment our efforts with more robust documentation.

File: index.html [hunk 4]
```diff
@@ -2,24 +2,13 @@
 <html lang="en">
     <head>
         <meta charset="UTF-8">
-        <link rel="stylesheet.css"/>
-        <title>gatehopper</title>
+        <link rel="stylesheet" type="text/css" href="stylesheet.css"/>
+        <title>Gatehopper</title>
```
Comment (rodneybarnes) [inline index.html]: There seems to be a lot of noise in this PR - not sure why these changes are showing up in this diff, as they're already on `main`. Is this something you would be able to investigate and remedy? That would make it easier to review, so we're not hunting through the diff to see what changes are actually being introduced.

File: src/objects/andGate.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import TwoInputGate from "./twoInputGate.js";
```
Comment (rodneybarnes) [inline src/objects/andGate.js]: We seem to have a lot of gate objects, would it be worth it to put them in their own folder inside our objects folder?

File: src/objects/gate.js (hunk 1 above)
Comment (VictorSohier) [reply src/objects/gate.js]: I looked up for a way to make abstract classes in javascript and because there is no official keyword to turn a class abstract, I looked it up and that was what I found.

File: index.html (hunk 4 above)
Comment (VictorSohier) [reply index.html]: It might be because I started it before the refactor and now we would be merging it in after the refactor. I might merge again once more to clean it up, especially now that I am using meld rather than a text editor to resolve conflicts.

File: src/objects/andGate.js (hunk 5 above)
Comment (VictorSohier) [reply src/objects/andGate.js]: Probably a good idea.

File: src/objects/xnorGate.js (hunk 2 above)
Comment (VictorSohier) [reply src/objects/xnorGate.js]: I wasn't paying attention to how I was exporting the gates. This will be fixed by removing the curly Brackets.

File: src/scenes/challengeMode.js (hunk 3 above)
Comment (VictorSohier) [reply src/scenes/challengeMode.js]: Resolved via a factory object

**Comment by VictorSohier** [comment]
Fixed some kind of weird overwriting issue with the cheatsheet, be it something I wrote myself or git not working the way I intended it to, it is now fixed.

File: src/objects/gateFactory.js [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+		    			this.width,
+		    			this.height)
+		    	});
+		    	resolve(true);
+	    	}));
+		}
+   		return promises;
+	}
+
+	create(gateType, input1, input2)
```
Comment (rodneybarnes) [inline src/objects/gateFactory.js:60]: I like the way you've set this up!

File: src/scenes/challengeMode.js [hunk 7]
```diff
@@ -1,15 +1,49 @@
...
+    			this.constant1 = this.gateFactory.create(GateTypeEnum.CONST, 0);
+    			this.constant2 = this.gateFactory.create(GateTypeEnum.CONST, 1);
+
+				this.gates.push(
+					this.gateFactory.create(
+						1, this.constant1
+					)
+				);
+
+    			for (let i = 2; i < 8; i++)
```
Comment (rodneybarnes) [inline src/scenes/challengeMode.js:31]: I'm assuming this is just for demonstration purposes?

**Review by rodneybarnes — APPROVED** [review]
Nicely done - looks like all of the gates show up. There are a couple points in the ticket's acceptance criteria I think we'll need to follow up on still, namely that we need a way for the player to input their answer for each logic gate, but I can create a new issue to cover that.

image
